### ISSUE TITLE
after wget and lynx, apt-cyg should fallback to curl too, if available

### ISSUE THREAD

**Issue by zaxebo1** [transcode-open/apt-cyg#99]
apt-cyg downloads packages using wget and falls back to lynx. I will request to kindly make apt-cyg fallback to "curl" too, if available.

even if curl is not available then it should fallback to the following script(which is pure bash)
http://superuser.com/a/496572

**Comment by keithdtyler** [comment]
`curl` is part of default Cygwin install and is required by other base tools, iirc, so it would be best if curl wasn't just the fallback, but the primary option.

    $ apt-cyg install nano
    Installing nano
    wget is not installed, using lynx as fallback
    /usr/bin/apt-cyg: line 128: lynx: command not found
    nano-2.6.2-1.tar.xz: FAILED

*alas*

**Comment by svnpenn** [comment]
https://github.com/svnpenn/sage/issues/16

**Comment by svnpenn** [comment]
@keithdtyler cURL is **not** part of default Cygwin install:

> category: Net

**Comment by svnpenn** [comment]
@keithdtyler cURL is **not** part of default Cygwin install:

> category: Net

Here are the default packages if you include dependencies:

~~~
_autorebase alternatives base-cygwin base-files bash bzip2 ca-certificates
coreutils cygutils cygwin dash diffutils editrights file findutils gawk getent
grep groff gzip hostname info ipc-utils less libargp libattr1 libblkid1 libbz2_1
libffi6 libgcc1 libgdbm4 libgmp10 libiconv libiconv2 libintl8 liblzma5 libmpfr4
libncursesw10 libopenssl100 libp11-kit0 libpcre1 libpipeline1 libpopt0
libpopt-common libreadline7 libsigsegv2 libsmartcols1 libssp0 libstdc++6
libtasn1_6 libuuid1 login man-db mintty ncurses openssl p11-kit p11-kit-trust
rebase run sed tar terminfo tzcode tzdata util-linux vim-minimal which xz zlib0
~~~

**Comment by native-api** [comment]
And what if I prefer `elinks`? Or `pwget`? Or `perl`/`python`/you-name-it?

`wget` is _the_ way. It has the fewest dependencies and, unlike reading from `/tcp/<whatever>`, it's sophisticated enough to handle things like unreliable transmission/throttling (not to mention FTP). So, I don't even see a use to fall back to _anything_ else.

**Comment by zaxebo1** [comment]
it is not about "preference", it is about "practicalitythat many installations (across globes) we have encountered has only curl installed, no wget/lynx. In that case if one has this apt-cyg script, then atleast apt-cyg should be able to download either fallback or atleast download wget using curl or **the pure bash code.** 
And then apt-cyg can continue with downloading packages with wget or lynx as it already does.

-----

That is, in this way one can be able to download most of the apt-cyg *pre-requisites* using apt-cyg script itself.

**Comment by heticor915** [comment]
I was testing this tool, because I was unaware of the requirements of apt-cyg I just install the base packages giving by cygwin in the  default installation, so the default package  curl is already installed but not lynx/wget so why no change this @zaxebo1 you could do a fork of this project and change the script

**Comment by svnpenn** [comment]
> the default package curl is already installed

@heticor915 no, no it is not.

**Comment by zaxebo1** [comment]
if we go by that above logic, as wget,lynx are not in base install, then we should drop the support of wget and lynx too from apt-cyg .
----
Somehow, with some choice installation of packages it happens that either curl or wget is there in most installations. This we have seen in many organizations installation. So, why not support that: if wget is not found, then apt-cyg can download wget using curl etc,for bootstrapping its pre-requisites

**Comment by svnpenn** [comment]
@zaxebo1 I already have a fix for this; you can bootstrap wget using C:\Windows\System32\ftp.exe, I just have not implemented it yet

**Comment by zaxebo1** [comment]
1) if MSWindows ftp.exe based solution  is implemented, then i am happy with that :+1:   (The only caution is that currently invoking Mswindows "ftp.exe" hangs the cygwin on my machine. ) 

2) If you implement ftp.exe based solution for installing wget, then the line to download "lynx -source rawgit.com/transcode-open/apt-cyg/master/apt-cyg > apt-cyg"  can also be alternatively put in terms of "ftp.exe ????" , then it will solve chicken and egg problem of having lynx through commandline for download apt-cyg itself. Hence, ftp.exe can help here too

**Comment by keithdtyler** [comment]
Guys, the point is that apt-cyg needs something to download with. It doesn't matter what it is. I thought curl was in base because I seem to recall it being used during cygwin  base installation and for a long time I cursed the installer because it would install curl and NOT wget, which I prefer. 

For me, the issue was that I ended up in a situation where I had *neither* wget or lynx. So in order to use apt-cyg, I had to go back to the GUI installer, which kind of feels like defeating the purpose. At least it was only one-time thing.

But bottom line, to be completely fair, a robust piece of software should not just go "Oop! Kablooey!" because something it expects to be there isn't there. It should either explain the problem or work around it. And right now, if you don't have wget or lynx, that's what apt-cyg does.

**Comment by native-api** [comment]
https://github.com/kou1okada/apt-cyg simply checks at startup if `tar`, `wget` and `awk` are installed and bails out with an error message if not. My vote goes there.

It's not much to ask from a user to install a few prerequisites one-time with the GUI installer, and `wget` is preferable to `curl` as I explained earlier.

**Comment by svnpenn** [comment]
Sage will now fall back to FTP in the event that wget is not found. I still need to tidy it up in that the server is currently hardcoded, but it should be working

https://github.com/svnpenn/sage/commit/535db64338ebcad9fa62f462a8c64fa2cacdd335

**Comment by keithdtyler** [comment]
Sorry I added a constructive comment to your code. Didn't mean to brutally assault your ego. Jeebus Crips.

**Comment by svnpenn** [comment]
https://github.com/transcode-open/apt-cyg/issues/103

**Comment by svnpenn** [comment]
https://github.com/transcode-open/apt-cyg/pull/105

**Comment by ivandavidov** [comment]
I vote for including ``curl`` as either primary download option, or as primary backup option after ``wget``. Otherwise, the current approach works fine and I'm OK with it.

**Comment by nhed** [comment]
I really think you should include curl.  I don't care if you include other fallbacks (wget->lynx->curl->/dev/tcp) - but you should do something before the world starts forking the heck out of this repo.  This is a great tool but its broken (i.e. when i instruct newbs on how to use it and if need to tell them to go back to setup - its a fail)

**Comment by RichardBronosky** [comment]
The maintainer doesn't care. I give up. Which fork should I use?

**Comment by keithdtyler** [comment]
He's still committing, but doesn't seem interested in working on any users' issues.

**Comment by RichardBronosky** [comment]
Here is how you can get `wget` if all you have is `curl`:

```
# check to see if you are running 64 bit cygwin
$ uname -a
CYGWIN_NT-10.0 WINDOWS-ABMESEI 2.6.0(0.304/5/3) 2016-08-31 14:32 x86_64 Cygwin

# if you are not using 64 bit, get http://www.cygwin.com/setup-x86.exe instead of...
$ curl -o cygwin-setup.exe http://www.cygwin.com/setup-x86_64.exe
$ chmod +x cygwin-setup.exe

# now you are ready to use it according to: https://cygwin.com/faq/faq.html#faq.setup.cli
$ cygwin-setup.exe --no-desktop --no-shortcuts --no-startmenu --quiet-mode --packages wget
```

Now you have `wget` and can use `apt-cyg`. But, it begs the question… now why do I need it anyway.

**Comment by nhed** [comment]
i agree - using setup in silent mode may be a better bigger picture solution for `apt-cyg` but you can develop it, and invest and maintainers still wont take your contrib, or even interact - so why bother

feel free to use my fork - i don't have any plans to remove it till `apt-cyg` is made to just work regardless of guidance people received on initial install of cygwin
